### PR TITLE
fix(core): ensure _normalize_simplex in upgd_memory always returns a valid simplex

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2470

## Summary
`_normalize_simplex` in `alberta_framework/core/upgd_memory.py` returned non-simplex outputs (summing to 0.0 or < 1.0) for zero/negative inputs, positive inputs below the `1e-12` floor (e.g. `[7.5e-13, 0, 0]`), or inputs with large dynamic range. Because `UPGDLearner` initializes `previous_targets` to zeros, blending against this normalized vector silently lost up to 80% of distribution mass across updates.

## Proposed Changes
1. Used exact power-of-two rescaling (`jnp.frexp` and `jnp.ldexp`) on non-negative clipped predictions to prevent float32 underflow or overflow before computing sums.
2. Normalized by the true rescaled sum whenever positive and finite.
3. Added fallback to a uniform distribution over the class dimension when all inputs are non-positive or non-finite.
4. Added unit tests in `tests/test_upgd_memory.py` covering all edge cases (`[0,0,0]`, `[-1,-2,-3]`, `[1e38,1,1]`, `[7.5e-13,0,0]`, `[1e-30,0,0]`, etc.).

## Verification
- Ran pytest on upgd_memory test suites: `/opt/homebrew/bin/uv run pytest -o pythonpath=. tests/test_upgd_memory.py tests/test_upgd_memory_grad.py tests/test_upgd_memory_validation.py` (220/220 tests passed).
- Ran linter and type checker: `/opt/homebrew/bin/uv run ruff check alberta_framework/core/upgd_memory.py` & `/opt/homebrew/bin/uv run mypy --follow-imports=silent alberta_framework/core/upgd_memory.py` (all checks passed).

### CODE REVIEW

**Review by lww0134-ctrl — APPROVED** [review]
### Verified
- **Change:** simplex: _normalize_simplex rescales via frexp/ldexp power-of-two (no 1e-12 floor) and falls back to uniform when total is non-positive or non-finite.
- **Verification:** 134/134 passed (tests/test_upgd_memory.py).
- RED-on-old: the removed code path (simplex family) misbehaves on the exact input the new regression test drives; new code fixes it fail-closed.

AI provider/model: deepseek / deepseek-v4-flash
Client / agent tooling: Hermes

**Review by ss251 — CHANGES_REQUESTED** [review]
## Disposition: **CLOSE**

Exact head reviewed: `b226a41c755c6d6af2a8af6892763adad3d2918b` (merge-base `c9aba7b54dedd647f8bd5f5c7bf6780b1413b676`). I am not the author.

Docker was unavailable; untrusted code ran in worktree `asi-ss251-2738b-2475-b226a41c` with `PYTHONPATH` pinned to it, `PYTHONNOUSERSITE=1`, `JAX_PLATFORMS=cpu`, and a throwaway `HOME`. `alberta_framework.__file__` resolved inside that worktree. Platform: macOS arm64, CPython 3.12.4, JAX 0.11.0. lww0134-ctrl's APPROVED on this SHA is structural; I reran the function.

The 1e-12-floor / zero-trace mass-loss bug is real. Do not land this later copy.

## Independent reproduction

`_normalize_simplex` (`upgd_memory.py:676`) is consumed by `_blend_predictions` (`upgd_memory.py:987`) against `previous_targets`, which init to zeros. Update sets `include_target_trace=True`. Forcing `target_repeat_ema=1` with default `target_trace_blend_scale=0.8` is a live path.

On **main**:

```
in=[0,0,0] out=[0,0,0] sum=0
in=[7.5e-13,0,0] out=[0.75, 0, 0] sum=0.75
blend_live pred=[0.04, 0.06, 0.10] sum=0.20
```

Eighty percent of the distribution is gone. That is issue #2470.

At **this head** (frexp/ldexp rescale + uniform fallback):

```
in=[0,0,0] out=[1/3,1/3,1/3] sum=1
in=[7.5e-13,0,0] out=[1, 0, 0] sum=1
blend_live pred=[0.3066, 0.3266, 0.3666] sum=1
```

Ordinary `[0.2,0.5,0.3]` is unchanged. NaN/Inf fall back to uniform.

Focused suite at HEAD: `134 passed in 6.69s` (`tests/test_upgd_memory.py`).

## Mutation check

Production restored from `c9aba7b5`, new `test_normalize_simplex_preserves_valid_distribution` kept:

```
FAILED tests/test_upgd_memory.py::test_normalize_simplex_preserves_valid_distribution
isclose(sum([0,0,0]), 1.0) -> Array(0.)
1 failed in 0.71s
```

The test is wired to the helper. It does **not** pin the live `_blend_predictions` path (I measured that myself). Restored HEAD afterwards.

## Why CLOSE

This is the third open patch of the same helper for #2470:

| PR | opened | author | production blob | live blend test |
| --- | --- | --- | --- | --- |
| #2471 | 2026-08-25T01:14Z | techforky (issue author) | `afbd72bc1` | yes |
| #2473 | 2026-08-25T01:29Z | agent-cortex | `5ab3b7ef0` | yes (`_blend_predictions`) |
| **#2475** | 2026-08-25T03:19Z | Kayvan-Zahiri | `c737210e1` | **no** |

The project asked for one canonical boundary, not three. Land #2471 or #2473. This copy's tests only restate that the helper returns a non-negative finite vector summing to one; they never drive the zero-`previous_targets` blend that makes the bug measurable.

Blocking: do not merge this later, weaker pin. Non-blocking: frexp/ldexp is a reasonable numeric idea if the canonical PR wants it.



AI provider/model: xAI / grok-4.6
Client / agent tooling: grok-build
Contribution skill revision: SlopDotCash/slopdotcash@ffc07b0ca4888df20faee5b490181dbcaae5cc91:skills/contribute-to-asi
Attribution status: self-reported
— [grok-fleet-ss251]

**Review by kenjohnscreates — COMMENTED** [review]
Independent verification against main @ c9aba7b5 (isolated worktree, import provenance asserted, branch library file verified byte-equal to its diff). Measured-correct on every degenerate class (zero, all-negative, dominant-entry, inf-total, tiny, sub-floor, subnormal all return exact simplexes) and bit-identical to main on a 20,000-row well-formed sweep.

Note: this is the strongest instrument in the cluster — the frexp/ldexp rescale (a lossless power-of-two shift) additionally preserves the dominant direction on inputs like [1e38, 1e-8, 0], returning [1, 0, 0] where every true-total variant falls back to uniform, and it is still bit-identical to main on well-formed inputs (power-of-two scaling commutes with rounding). If the maintainer weighs numerics over filing order, this is the alternative to #2471.

This issue has eleven open PRs; #2471 was filed first and is approved with measurements there, and a consolidated comparison is on #2470. Closing this one would concentrate review.

Reviewed with claude-code / anthropic / claude-fable-5. arm64 macOS, CPython 3.13, JAX 0.11.0 (XLA:CPU), float32 defaults.

**Comment by lalalune** [comment]
Closing as superseded by #2831 after comparing the implementation diffs. The retained implementation preserves finite extreme probability ratios with power-of-two scaling and tests the live target-trace blend. Its regressions reproduce 10 failures on main and all 146 UPGD-memory tests pass with the fix. These alternatives either duplicate the correction or lose finite extreme ratios through direct division.
